### PR TITLE
Factor IntervalScan out of ServiceDispatchHandler peer reap

### PR DESCRIPTION
**Pull request by jcorbin**
Factor IntervalScan out of ServiceDispatchHandler peer reaping

r @kriskowal @Raynos

### CODE REVIEW

File: service-proxy.js [hunk 1]
```diff
@@ -112,8 +113,25 @@ function ServiceDispatchHandler(options) {
...
+            return peersToReap;
+        },
+        self.channel.timers
+    );
+    self.peerReaper.runBeginEvent.on(function onPeerReapBegin(run) {
+        self.logger.info('reaping dead peers', self.extendLogInfo({
+            numPeersToReap: run.keys.length
+        }));
+    });
+    self.peerReaper.start(); // TODO: should IntervalScan be self-starting?
```
Comment (kriskowal) [inline service-proxy.js]: ServiceProxy itself starts on construction and stops on destruction. Perhaps, having start/stop methods should be a thing you build so you can pause/resume. destroy should stop regardless.

File: service-proxy.js [hunk 2]
```diff
@@ -112,8 +113,25 @@ function ServiceDispatchHandler(options) {
...
+            self.knownPeers = Object.create(null);
+            return peersToReap;
+        },
+        self.channel.timers
+    );
+    self.peerReaper.runBeginEvent.on(function onPeerReapBegin(run) {
+        self.logger.info('reaping dead peers', self.extendLogInfo({
+            numPeersToReap: run.keys.length
+        }));
+    });
```
Comment (kriskowal) [inline service-proxy.js]: Usage looks good.

File: service-proxy.js [hunk 3]
```diff
@@ -112,8 +110,53 @@ function ServiceDispatchHandler(options) {
...
+                    delete self.channel.subChannels[serviceName];
+                    self.rateLimiter.removeServiceCounter(serviceName);
+                    self.rateLimiter.removeKillSwitchCounter(serviceName);
+                }
+            }
+        },
+        // TODO: would be convenient to be able to pass an object or a function
+        // in cases like this
+        function getExitServices() {
+            return self.exitServices;
```
Comment (kriskowal) [inline service-proxy.js]: Consider capturing a snapshot.

Comment (jcorbin) [inline service-proxy.js]: Added a guard to IntervalScanRun#nextItem to skip undefined entries, I'd rather not introduce a copy into places like this where they didn't exist before; also I actively want to get a, maybe newer, lastRefresh by the time I get around to calling `maybePurgeEachService`

**Comment by jcorbin** [comment]
I'm not 100% happy with the new constructor bloat, how about one or more .setupX methods?

**Comment by jcorbin** [comment]
Punting the other two use cases to future PRs.

Also deferring any new Features like observability to future work

File: service-proxy.js [hunk 4]
```diff
@@ -112,8 +112,26 @@ function ServiceDispatchHandler(options) {
     self.peersToReap = Object.create(null);
     self.knownPeers = Object.create(null);
 
-    self.reapPeersTimer = null;
-    self.reapPeersPeriod = options.reapPeersPeriod || DEFAULT_REAP_PEERS_PERIOD;
+    self.peerReaper = new IntervalScan(
+        'peer-reap',
```
Comment (Raynos) [inline service-proxy.js]: You are my friend.

File: lib/interval-scan.js [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    if (self.callback) {
+        self.callback();
+        self.callback = null;
+    }
+};
+
+IntervalScanRun.prototype.clearImmed =
+function clearImmed() {
```
Comment (kriskowal) [inline lib/interval-scan.js]: typo: misspelling of Immediate

**Comment by Raynos** [comment]
#isvalid

File: lib/interval-scan.js (hunk 5 above)
Comment (Raynos) [inline lib/interval-scan.js]: 80 characters.

Comment (Raynos) [inline lib/interval-scan.js]: Immediate.

**Comment by kriskowal** [comment]
:ok:

**Pull request merged**